### ISSUE TITLE
SET dismiss modal regression in latest version (7.12.0)

### ISSUE THREAD

**Issue by nelsonprsousa** [wix/react-native-navigation#7070]
## 🐛 Bug Report
While upgrading from 7.8.4 to the latest version 7.12.0, shared element transition no longer works as expected while dismissing modals. Also tested with master (cef992918cce652427d85d6a52e8d9ded65b95c6), same problem.

### Have you read the [Contributing Guidelines on issues](https://wix.github.io/react-native-navigation/docs/meta-contributing/)?
Yes.

## To Reproduce

1. Clone [this repo](https://github.com/nelsonprsousa/rnn-testing/tree/set-regression)
2. Use `set-regression` branch
3. yarn
4. yarn ios / yarn android
5. The last commit is the upgrade to 7.12.0. You can just drop the last commit to go back to 7.8.4, without this problem (reinstall pods and gradlew clean for Android).

## Expected behavior
The background screen should appear as before when we dismiss the modal.

## Actual Behavior
Looks like the screen just appears once the SET is finished.

## Your Environment

- React Native Navigation version: 7.12.0
- React Native version: 0.63.4
- Platform(s) (iOS, Android, or both?): both
- Device info (Simulator/Device? OS version? Debug/Release?): Android and iOS

## Reproducible Demo

7.8.4:

https://user-images.githubusercontent.com/8364345/113493667-6a781200-94d9-11eb-8a50-d68d7b6aaeca.mov

7.12.0:

https://user-images.githubusercontent.com/8364345/113493704-c347aa80-94d9-11eb-8568-974f0d3cffbd.mov

**Comment by nelsonprsousa** [comment]
Also noticed that on Android the SET while showing the modal is no longer working with the latest version.

**Comment by nelsonprsousa** [comment]
Hey @guyca and @swabbass 
I don't think the PR that was merged completely solves this issue. On dismiss model, looks like the background isn't correctly showed. I tried to replicate it with slow motion here (using current snapshot: 7.13.0-snapshot.1515):

https://user-images.githubusercontent.com/8364345/114719786-54671e80-9d2f-11eb-8a6f-3d057b014378.mov

**Comment by guyca** [comment]
@nelsonprsousa The PR fixed an issue on Android while the video shows iOS 🤔

**Comment by nelsonprsousa** [comment]
> @nelsonprsousa The PR fixed an issue on Android while the video shows iOS 🤔
Check the last video on the original video. It's both for Android and iOS.

Uploading Screen Recording 2021-04-14 at 16.36.48.mov…

**Comment by swabbass** [comment]
> > @nelsonprsousa The PR fixed an issue on Android while the video shows iOS 🤔
> 
> Check the last video on the issue. It's both for Android and iOS.
> 
>  screen-recording-2021-04-14-at-163648_uZA60di7_gffr.mov

@nelsonprsousa thx :]

As we talked about in the discord this is another issue, here the issue is the order of the fading views, the shared one, disappears as expected but the bg might have something. 

So, you can open a new one, and soon we will get to it, meanwhile you can try to understand what is wrong here since it is happening on ios and android.

**Comment by guyca** [comment]
@nelsonprsousa If the issue happens on both platforms, It's quite possible that it's a JS issue on your end.

This use case seems to work on both Android and iOS in the playground app following the fix by Ward.

**Comment by guyca** [comment]
Tested the repo provided with the latest snapshot - doesn't seem like the bug is fixed. At least not on Android.

**Comment by swabbass** [comment]
> Tested the repo provided with the latest snapshot - doesn't seem like the bug is fixed. At least not on Android.

@guyca can you elaborate on which bug?
the one described above is fixed (the first one),  where the animations won't start, the second case is different I think.

**Comment by nelsonprsousa** [comment]
> > Tested the repo provided with the latest snapshot - doesn't seem like the bug is fixed. At least not on Android.
> 
> @guyca can you elaborate on which bug?
> the one described above is fixed (the first one), where the animations won't start, the second case is different I think.

I didn't understand  what bug was solved, but the one reported in this issue is this one:

>  ## Expected behavior
> The background screen should appear as before when we dismiss the modal.
> 
> ## Actual Behavior
> Looks like the screen just appears once the SET is finished.

It's also on the title: SET dismiss modal regression in latest version

I am not sure what to do 🤔

**Comment by nelsonprsousa** [comment]
If you can I can re-create the issue, but I think I'll copy paste this one?!

**Comment by swabbass** [comment]
> If you want I can re-create the issue, but I think I'll copy paste this one?!

Ohh, I see, no need for a new issue, the screenshots, have 2 issues, thought that the reported issue was just about the animations that did not play...So I'm really sorry for being confused here, I'll take that and add another PR soon :]

**Comment by swabbass** [comment]
@nelsonprsousa 
Ok so here is the thing, aside from the first issue where the animation was not running on card click, there is another issue about fade in/out the modals (a parent who is holding the shared element ) once animation ended:

I reproduced the same behaviour that in the playground (Car dealer example) with:

```javascript
// Contsatnts.ts
export function buildSharedElementDefaultDismiss(car: CarItem): AnimationOptions {
  return {
    showModal: {
      sharedElementTransitions: [
        {
          fromId: `image${car.id}`,
          toId: `image${car.id}Dest`,
          duration: SET_DURATION,
          interpolation: { type: 'spring', ...SPRING_CONFIG },
        },
        {
          fromId: `title${car.id}`,
          toId: `title${car.id}Dest`,
          duration: SET_DURATION,
          interpolation: { type: 'spring', ...SPRING_CONFIG },
        },
      ],
    },
  };
}
```

from here I see that if you want the modals (parents) to have animation when `show/Dismiss` modal, you need to override the animation options to have the desired behaviour, something like this will do the job in the project you mentioned:

![Screen Shot 2021-04-19 at 1 12 12](https://user-images.githubusercontent.com/7227793/115162684-8bca2800-a0ad-11eb-961f-68b65461cf11.png)

with this little change we got the issue fixed like this:

https://user-images.githubusercontent.com/7227793/115162716-ac927d80-a0ad-11eb-8a1f-da3030c54eba.mov

It seems that you relied on the idea that is, if in showing modal we have fade-in animation, then we got to have a fade-out as default for dismissing modal, which is missing, there is no default fade animation both ways, as for today.

I tried to make little changes in the RNN native side, in order to implement both ways of the default fade animations, and I got this:

https://user-images.githubusercontent.com/7227793/115162856-8faa7a00-a0ae-11eb-8ac2-bf82ee298431.mov

so this works, but something seems not right for me.
I want to share it with you @nelsonprsousa , @guyca , @yogevbd  and see what do you think?, here are my thoughts:

-The default fade animation has fixed and independent duration, which means whatever animation running in the command (`showModal`, `dismissModal`) that is not fade, will be out of sync with the other animations, especially if the user expecting the animations to share the same duration or some reasonable amount of duration that matches their need.

- If we want to fully support default fade animations that is cool, but let's think about what animation element in `showModal`, for example, should override the default fade animation duration without it (alpha) being declared?, or, what about keeping it as is, hardcoded and improve the docs and explain that we got this as default?,  Maybe the user needs to define the desired animation behaviour without relying on RNN to provide some default behaviour since the API is customizable for this kind of stuff.

**Comment by nelsonprsousa** [comment]
Thanks for your explanation @swabbass ! I might be missing something, but what "duration" prop means in the "alpha" object then? 🤔

**Comment by guyca** [comment]
@swabbass Great job debugging this! Whenever SET is involved, fade animations should be used by default. We're already using the fade animation for SET in push/pop, the same should be done for dismiss/show modals.

**Comment by swabbass** [comment]
> Thanks for your explanation @swabbass ! I might be missing something, but what "duration" prop means in the "alpha" object then? 🤔

actually, you are not missing anything, it was a regression, it was playing, and now it is fixed, sorry for the inconvenience :].
